{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genotype data (SNP) QC pipeline\n",
    "\n",
    "SNP 데이터를 preprocessing하는 과정의 pipeline입니다.\n",
    "\n",
    "\n",
    "필요한 소프트웨어는 PLINK1.9버전(https://www.cog-genomics.org/plink/) 입니다.\n",
    "\n",
    "폴더 안에는 PLINK와 preprocess할 genotype data가 binary plink format(bed,bim,fam)의 확장자를 가진 같은 이름으로 존재해야 합니다.\n",
    "\n",
    "\n",
    "Phenotype은 넣지 않아도 돌아가게 만들었습니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "이 곳의 파라미터를 바꿈으로써 preprocessing의 여러 threshold를 바꿀 수 있습니다.\n",
    "최종결과는 'ABCD_finished.(fbed,bim,fam)'파일로 나오게 됩니다.\n",
    "'''\n",
    "#작업할 파일 공간(이 안에 genotype data와 plink 소프트웨어를 넣어두어야 합니다)\n",
    "Filepath = '/home/hjd/Desktop/PLINK/test'\n",
    "\n",
    "#Genotype data의 파일명입니다.\n",
    "Filename = 'ABCD_release_3.0_QCed'\n",
    "\n",
    "#제외해야 할 목록이 있다면 그 목록이 적힌 텍스트파일의 이름을 쓰면 됩니다. 물론 이 파일도 filepath 안에 넣어놔야 합니다.\n",
    "Bad_batch_file = 'batch_bad.txt'\n",
    "\n",
    "#Phenotype의 정보가 들어있는 파일이름입니다. 없으면 놔두셔도 무방합니다.\n",
    "#파일의 첫column은 Family ID, 두번째column은 Individual ID, 그리고 3번째 column이 phenotype(trait)입니다.\n",
    "Phenotype_file = None #'qtrait.txt'\n",
    "\n",
    "#--geno,--mind,MAF,HWE의 threshold를 설정합니다.\n",
    "Genotype_miss = None\n",
    "Individual_miss = None\n",
    "MAF = None\n",
    "HWE = None\n",
    "\n",
    "#몇번부터 몇번 염색체까지 쓸 것인지 설정합니다.\n",
    "Chromosome = None #[,]\n",
    "\n",
    "#Prune에 필요한 수치를 설정합니다.\n",
    "Prune_window_size = None\n",
    "Prune_sliding = None\n",
    "Prune_r2 = None\n",
    "\n",
    "#Heterozygosity에서 몇 standard deviation이상을 제외할건지 threshold를 설정합니다.\n",
    "Heterozygosity_SD = 4\n",
    "\n",
    "#Relatedness 분석에서 최솟값과 최댓값을 설정합니다.\n",
    "Relatedness_min = 0.4\n",
    "Relatedness_max = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#기본세팅입니다. 값을 입력하지 않으면 이 값을 바탕으로 진행됩니다.\n",
    "\n",
    "config = {\n",
    "    'Filepath' : None,\n",
    "    'Filename' : None,\n",
    "    'Bad_batch_file' : None,\n",
    "    'Phenotype_file' : None,\n",
    "    'Genotype_miss' : 0.05,\n",
    "    'Individual_miss' : 0.05,\n",
    "    'Minor allele frequency' : 0.01,\n",
    "    'Chromosome' : [1,22],\n",
    "    'Hardy weinberg equil' : 1e-6,\n",
    "    'Prune' : {\n",
    "        'window_size': 200,\n",
    "        'sliding': 50,\n",
    "        'r2': 0.25,\n",
    "    },\n",
    "    'Heterozygosity(SD)' : 3,\n",
    "    'Relatedness' : {\n",
    "        'min' : None,\n",
    "        'max' : None,\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=================START Preprocessing!======================\n",
      "\n",
      "=================1.Bad batch removing======================\n",
      "Bad batch file detected! Processing...\n",
      "=================1.Bad batch removed======================\n",
      "\n",
      "=================1.Checking phenotype======================\n",
      "No phenotype file detected. Run preprocess without phenotype.\n",
      "===================Step 1 completed=======================\n",
      "\n",
      "======2.Start excluding missing snps(--geno), threshold = 0.05======\n",
      "===================Step 2 completed=======================\n",
      "\n",
      "======3.Start excluding missing individuals(--mind), threshold = 0.05=====\n",
      "===================Step 3 completed=======================\n",
      "\n",
      "=================4.Imputing sex======================\n",
      "===================Step 4 completed=======================\n",
      "\n",
      "======5.Choosing chromosome : 1 to 22======\n",
      "===================Step 5 completed=======================\n",
      "\n",
      "======6.Minor allele frequency, threshold = 0.01======\n",
      "===================Step 6 completed=======================\n",
      "\n",
      "======7.Hardy weinberg equilibrium, threshold = 1e-06======\n",
      "===================Step 7 completed=======================\n",
      "\n",
      "====================8.Start pruning... ===================\n",
      "window size = 200\n",
      "sliding = 50\n",
      "r2 = 0.25\n",
      "======Making pruned data... ======\n",
      "===================Step 8 completed=======================\n",
      "\n",
      "===========9.Start heterozygosity calculation ============\n",
      "======Exclude 4SD away from mean value ======\n",
      "======het_remove file has been made!======\n",
      "===================Step 9 completed=======================\n",
      "\n",
      "======10.Calculating Relatedness...this takes time, be patient :) ======\n",
      "====== min : 0.4, max : None======\n",
      "========Comparing individuals========\n",
      "========Saving low call individual text file========\n",
      "\n",
      "===================Step 10 completed=======================\n",
      "\n",
      "=============Genotype preprocessing done!!!==================\n"
     ]
    }
   ],
   "source": [
    "import os, sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "#Config\n",
    "config['Filepath'] = Filepath\n",
    "config['Filename'] = Filename\n",
    "config['Bad_batch_file'] = Bad_batch_file\n",
    "config['Phenotype_file'] = Phenotype_file\n",
    "config['Genotype_miss'] = config['Genotype_miss'] if Genotype_miss == None else Genotype_miss\n",
    "config['Individual_miss'] = config['Individual_miss'] if Individual_miss == None else Individual_miss\n",
    "config['Minor allele frequency'] = config['Minor allele frequency'] if MAF == None else MAF\n",
    "config['Chromosome'] = config['Chromosome'] if Chromosome == None else Chromosome\n",
    "config['Prune']['window_size'] = config['Prune']['window_size'] if Prune_window_size == None else Prune_window_size\n",
    "config['Prune']['sliding'] = config['Prune']['sliding'] if Prune_sliding == None else Prune_sliding\n",
    "config['Prune']['r2'] = config['Prune']['r2'] if Prune_r2 == None else Prune_r2\n",
    "config['Heterozygosity(SD)'] = config['Heterozygosity(SD)'] if Heterozygosity_SD == None else Heterozygosity_SD\n",
    "config['Relatedness']['min'] = Relatedness_min\n",
    "config['Relatedness']['max'] = Relatedness_max\n",
    "\n",
    "os.chdir(config['Filepath'])\n",
    "os.mkdir(\"Final_results\")\n",
    "#Save config\n",
    "import json\n",
    "with open('Final_results/Used_parameters.json', 'w') as f:\n",
    "        json.dump(config, f,indent = 0)\n",
    "\n",
    "print('=================START Preprocessing!======================\\n')\n",
    "print('=================1.Bad batch removing======================')\n",
    "#Exclude bad batch\n",
    "if config['Bad_batch_file'] != None:\n",
    "    print('Bad batch file detected! Processing...')\n",
    "    os.mkdir('1.Bad_batch_removed')\n",
    "    os.system('./plink --bfile ABCD_release_3.0_QCed --remove {} --out 1.Bad_batch_removed/ABCD_excluded --make-bed'.format(config['Bad_batch_file']))\n",
    "else:\n",
    "    print('No bad batch file detected. Anyway, New file will be made!')\n",
    "    os.mkdir('1.Bad_batch_removed')\n",
    "    os.system('./plink --bfile ABCD_release_3.0_QCed --out 1.Bad_batch_removed/ABCD_excluded --make-bed'.format(config['Bad_batch_file']))\n",
    "\n",
    "print('=================1.Bad batch removed======================\\n')\n",
    "\n",
    "print('=================1.Checking phenotype======================')\n",
    "if config['Phenotype_file'] != None:\n",
    "    if os.path.isfile(config['Phenotype_file']):\n",
    "        os.mkdir('Phenotyped')\n",
    "        print('Phenotype file detected!! Processing...')\n",
    "        os.system('./plink --bfile 1.Bad_batch_removed/ABCD_excluded --pheno {} --out Phenotyped/ABCD_phenotyped --make-bed'.format(config['Phenotype_file']))\n",
    "    else:\n",
    "        print('No phenotype file detected. Run preprocess without phenotype.')\n",
    "        pass\n",
    "else: print('No phenotype file detected. Run preprocess without phenotype.')\n",
    "print('===================Step 1 completed=======================\\n')\n",
    "\n",
    "#Exclude missing genotype\n",
    "print('======2.Start excluding missing snps(--geno), threshold = {}======'.format(config['Genotype_miss']))\n",
    "os.mkdir('2.Missing_genotype')\n",
    "if config['Phenotype_file'] != None:\n",
    "    os.system('./plink --bfile Phenotyped/ABCD_phenotyped --geno {} --out 2.Missing_genotype/ABCD_geno --make-bed'.format(config['Genotype_miss']))\n",
    "else:\n",
    "    os.system('./plink --bfile 1.Bad_batch_removed/ABCD_excluded --geno {} --out 2.Missing_genotype/ABCD_geno --make-bed'.format(config['Genotype_miss']))\n",
    "print('===================Step 2 completed=======================\\n')\n",
    "    \n",
    "print('======3.Start excluding missing individuals(--mind), threshold = {}====='.format(config['Individual_miss']))\n",
    "#Exclude missing individuals\n",
    "os.mkdir('3.Missing_individuals')\n",
    "os.system('./plink --bfile 2.Missing_genotype/ABCD_geno --mind {} --out 3.Missing_individuals/ABCD_mind --make-bed'.format(config['Individual_miss']))\n",
    "print('===================Step 3 completed=======================\\n')\n",
    "\n",
    "#Impute sex\n",
    "print('=================4.Imputing sex======================')\n",
    "os.mkdir('4.Impute_sex')\n",
    "os.system('./plink --bfile 3.Missing_individuals/ABCD_mind --impute-sex --out 4.Impute_sex/ABCD_sex --make-bed')\n",
    "print('===================Step 4 completed=======================\\n')\n",
    "\n",
    "#Choose which chromosome to use\n",
    "print('======5.Choosing chromosome : {} to {}======'.format(config['Chromosome'][0],\n",
    "                                                         config['Chromosome'][1]))\n",
    "os.mkdir('5.Choose_chromosome')\n",
    "os.system('./plink --bfile 4.Impute_sex/ABCD_sex --chr {}-{} --out 5.Choose_chromosome/ABCD_chr --make-bed'.format(config['Chromosome'][0],\n",
    "                                                                                                                  config['Chromosome'][1]))\n",
    "print('===================Step 5 completed=======================\\n')\n",
    "\n",
    "#Minor Allele Frequency\n",
    "print('======6.Minor allele frequency, threshold = {}======'.format(config['Minor allele frequency']))\n",
    "os.mkdir('6.Minor_allele_frequency')\n",
    "os.system('./plink --bfile 5.Choose_chromosome/ABCD_chr --maf {} --out 6.Minor_allele_frequency/ABCD_maf --make-bed'.format(config['Minor allele frequency']))\n",
    "print('===================Step 6 completed=======================\\n')\n",
    "\n",
    "#Hardy Weinberg Equilibrium\n",
    "print('======7.Hardy weinberg equilibrium, threshold = {}======'.format(config['Hardy weinberg equil']))\n",
    "os.mkdir('7.HWE')\n",
    "os.system('./plink --bfile 6.Minor_allele_frequency/ABCD_maf --hwe {} --out 7.HWE/ABCD_hwe --make-bed'.format(config['Hardy weinberg equil']))\n",
    "print('===================Step 7 completed=======================\\n')\n",
    "\n",
    "#Pruning\n",
    "print('====================8.Start pruning... ===================')\n",
    "print('window size = {}'.format(config['Prune']['window_size']))\n",
    "print('sliding = {}'.format(config['Prune']['sliding']))\n",
    "print('r2 = {}'.format(config['Prune']['r2']))\n",
    "os.mkdir('8.Pruning')\n",
    "os.system('./plink --bfile 7.HWE/ABCD_hwe --indep-pairwise {} {} {} --out 8.Pruning/ABCD_pruned'.format(config['Prune']['window_size'],\n",
    "                                                                                       config['Prune']['sliding'],\n",
    "                                                                                       config['Prune']['r2']))\n",
    "\n",
    "#Make pruned dataset\n",
    "print('======Making pruned data... ======')\n",
    "os.system('./plink --bfile 7.HWE/ABCD_hwe --extract 8.Pruning/ABCD_pruned.prune.in --out 8.Pruning/ABCD_pruned --make-bed')\n",
    "print('===================Step 8 completed=======================\\n')\n",
    "\n",
    "print('===========9.Start heterozygosity calculation ============')\n",
    "#Heterozygosity with pruned dataset\n",
    "os.mkdir('9.Heterozygosity')\n",
    "os.system('./plink --bfile 8.Pruning/ABCD_pruned --het --out 9.Heterozygosity/ABCD_het')\n",
    "\n",
    "#Make removal list with heterozygosity calculation\n",
    "ABCD_het = pd.read_csv('9.Heterozygosity/ABCD_het.het', sep = ' ', low_memory = False,\n",
    "                       skipinitialspace=True)\n",
    "\n",
    "#Calculating heterozygosity\n",
    "ABCD_het['het'] = (ABCD_het['N(NM)']-ABCD_het['O(HOM)'])/ABCD_het['N(NM)']\n",
    "mean = np.mean(ABCD_het['het'])\n",
    "std = np.std(ABCD_het['het'])\n",
    "print('======Exclude {}SD away from mean value ======'.format(config['Heterozygosity(SD)']))\n",
    "over_out = ABCD_het.loc[ABCD_het['het'] > mean+int('{}'.format(config['Heterozygosity(SD)']))*std,:]\n",
    "under_out = ABCD_het.loc[ABCD_het['het'] < mean-int('{}'.format(config['Heterozygosity(SD)']))*std, :]\n",
    "remove = pd.concat([over_out,under_out], axis=0)\n",
    "remove.to_csv('het_remove.txt', sep = ' ', header=None, index = False)\n",
    "print('======het_remove file has been made!======')\n",
    "\n",
    "os.system('./plink --bfile 7.HWE/ABCD_hwe --remove het_remove.txt --out 9.Heterozygosity/ABCD_het_removed --make-bed;')\n",
    "print('===================Step 9 completed=======================\\n')\n",
    "\n",
    "print('======10.Calculating Relatedness...this takes time, be patient :) ======')\n",
    "print('====== min : {}, max : {}======'.format(config['Relatedness']['min'],config['Relatedness']['max']))\n",
    "os.mkdir('10.Relatedness')\n",
    "if config['Relatedness']['min'] != None and config['Relatedness']['max'] == None:\n",
    "    os.system('./plink --bfile 8.Pruning/ABCD_pruned --genome --min {} --out 10.Relatedness/ABCD_genome'.format(config['Relatedness']['min']))\n",
    "if config['Relatedness']['min'] == None and config['Relatedness']['max'] != None:\n",
    "    os.system('./plink --bfile 8.Pruning/ABCD_pruned --genome --max {} --out 10.Relatedness/ABCD_genome'.format(config['Relatedness']['max']))\n",
    "if config['Relatedness']['min'] != None and config['Relatedness']['max'] != None:\n",
    "    os.system('./plink --bfile 8.Pruning/ABCD_pruned --genome --min {} --max {} --out 10.Relatedness/ABCD_genome'.format(config['Relatedness']['min'],\n",
    "                                                                                                         config['Relatedness']['max']))\n",
    "if config['Relatedness']['min'] == None and config['Relatedness']['max'] == None:\n",
    "    os.system('./plink --bfile 8.Pruning/ABCD_pruned --genome --out 10.Relatedness/ABCD_genome')\n",
    "\n",
    "ABCD_genome = pd.read_csv('10.Relatedness/ABCD_genome.genome', sep = ' ', low_memory = False,\n",
    "                       skipinitialspace=True)\n",
    "\n",
    "#get missing rate information\n",
    "os.system('./plink --bfile 1.Bad_batch_removed/ABCD_excluded --missing --out 10.Relatedness/ABCD_miss')\n",
    "miss = pd.read_csv('10.Relatedness/ABCD_miss.imiss', sep = ' ', low_memory = False,\n",
    "                       skipinitialspace=True)\n",
    "\n",
    "print('========Comparing individuals========')\n",
    "#Compare individuals missing rate and leave only good one for each pairs\n",
    "low_call_list = []\n",
    "low_call_val = []\n",
    "for i in range(ABCD_genome.shape[0]):\n",
    "    f1 = ABCD_genome.loc[i,'FID1']\n",
    "    p1 = ABCD_genome.loc[i,'IID1']\n",
    "    f2 = ABCD_genome.loc[i,'FID2']\n",
    "    p2 = ABCD_genome.loc[i,'IID2']\n",
    "    c1 = float(miss.loc[miss['IID'] == p1, 'F_MISS'])\n",
    "    c2 = float(miss.loc[miss['IID'] == p2, 'F_MISS'])\n",
    "    if c1>c2:\n",
    "        low_call_list.append([f1,p1])\n",
    "        low_call_val.append([c1,c2])\n",
    "    elif c2>c1:\n",
    "        low_call_list.append([f2,p2])\n",
    "        low_call_val.append([c1,c2])\n",
    "    else: continue\n",
    "        \n",
    "#Save low call individual list\n",
    "print('========Saving low call individual text file========\\n')\n",
    "f = open(\"low_call.txt\", 'w')\n",
    "for i in range(len(low_call_list)):\n",
    "    a1 = str(low_call_list[i]).replace(\"['\",'')\n",
    "    a2 = a1.replace(\"']\",'')\n",
    "    a3 = a2.replace(\"', '\",' ')\n",
    "    data = a3 + '\\n'\n",
    "    f.write(data)\n",
    "f.close()\n",
    "print('===================Step 10 completed=======================\\n')\n",
    "#Exclude low call list from data\n",
    "os.system('./plink --bfile 9.Heterozygosity/ABCD_het_removed --remove low_call.txt --out Final_results/ABCD_QC_finished --make-bed')\n",
    "print(\"=============Genotype preprocessing done!!!==================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
